### ISSUE TITLE
Updates to services (selector) being ignored

### ISSUE THREAD

**Issue by lostllama** [datawire/ambassador#848]
**Describe the bug**
We're using blue/green deployments and effect the changeover by updating the selector of the two Ambassador services.

We've found that, although it sometimes works, often this change will be ignored. To test this, I added the commit id to a status API endpoint and can see that the swap doesn't happen (or worse, since we're using two Ambassador instances each with their own IDs, only one updates). After waiting about 10 minutes, no change is visible.

**To Reproduce**
1. Create two discernibly different API deployments, A and B, with different labels (e.g. API=A, API=B).
2. Create a single Ambassador service to point at A's label.
3. Access the API and confirm that it's A.
4. Edit the Ambassador service (kubebtl edit service ~~~~) to point at B's label.
5. Access the API and find it's still A. Repeat after 10 minutes. Still A.

**Expected behavior**
I expect Ambassador to update to point at the other deployment.

**Versions (please complete the following information):**
 - Ambassador: 0.40.0
 - GKE v1.9.7-gke.5

**Additional context**

If I try to change the values in the two files, and then I do nothing more than scale the Ambassador deployments in to 0 replicas, and back out, it will effect the change, so our config files are definitely correct.

I've found a workaround which is to add a dummy header which contains a 5 character alphanumeric string generated by Helm:

  add_request_headers:
      x-test-data: "{{ randAlphaNum 5 }}"

Adding this header causes Ambassador to update each time (and in under 30 seconds).

**Comment by nbkrause** [comment]
Could you paste the yaml of the service you are editing?

**Comment by lostllama** [comment]
With my exact configuration, it happens every time. With this one, which is very similar to my own (this is basically a compiled version of the Helm template I'm using, with the names changed to protect the innocent), it happens maybe 1/10 times where it just gets stuck on one.

I just keep swapping and eventually it gets stuck. I've got two deployments here so that it's obvious which is being hit: one is Apache and the other is Nginx.

    apiVersion: v1
    kind: ServiceAccount
    metadata:
      name: ambassador-release
      namespace: sandbox
    ---
    apiVersion: rbac.authorization.k8s.io/v1beta1
    kind: ClusterRole
    metadata:
      name: ambassador-release
      namespace: sandbox
    rules:
    - apiGroups: [""]
      resources:
      - services
      verbs: ["get", "list", "watch"]
    - apiGroups: [""]
      resources:
      - configmaps
      verbs: ["create", "update", "patch", "get", "list", "watch"]
    - apiGroups: [""]
      resources:
      - secrets
      verbs: ["get", "list", "watch"]
    ---
    apiVersion: rbac.authorization.k8s.io/v1beta1
    kind: ClusterRoleBinding
    metadata:
      name: ambassador-release
    roleRef:
      apiGroup: rbac.authorization.k8s.io
      kind: ClusterRole
      name: ambassador-release
    subjects:
    - kind: ServiceAccount
      name: ambassador-release
      namespace: sandbox
    ---
    apiVersion: v1
    kind: Service
    metadata:
      labels:
        env: "sandbox"
        type: ambassador
      name: ambassador-release-loadbalancer
      namespace: sandbox
    spec:
      externalTrafficPolicy: Cluster
      ports:
      - port: 80
        protocol: TCP
        name: web-http
        targetPort: 80

      selector:
        service: ambassador-release-service
      sessionAffinity: None
      type: LoadBalancer
    ---
    apiVersion: v1
    kind: Service
    metadata:
      labels:
        service: "ambassador-release-admin"
      name: ambassador-release-admin
      namespace: sandbox
    spec:
      type: NodePort
      ports:
      - name: "ambassador-release-admin"
        port: 8877
        targetPort: 8877
      selector:
        service: "ambassador-release-ambassador"
    ---	
    apiVersion: extensions/v1beta1
    kind: Deployment
    metadata:
      name: ambassador-release
      namespace: sandbox
    spec:
      replicas: 1
      template:
        metadata:
          annotations:
            sidecar.istio.io/inject: "false"
          labels:
            service: ambassador-release-service
        spec:
          serviceAccountName: ambassador-release
          containers:
          - name: ambassador
            image: quay.io/datawire/ambassador:0.40.0
            resources:
              limits:
                cpu: 1
                memory: 400Mi
              requests:
                cpu: 200m
                memory: 100Mi
            env:
            - name: AMBASSADOR_ID
              value: ambassador-release
            - name: AMBASSADOR_NAMESPACE
              valueFrom:
                fieldRef:
                  fieldPath: metadata.namespace
            livenessProbe:
              httpGet:
                path: /ambassador/v0/check_alive
                port: 8877
              initialDelaySeconds: 30
              periodSeconds: 3
            readinessProbe:
              httpGet:
                path: /ambassador/v0/check_ready
                port: 8877
              initialDelaySeconds: 30
              periodSeconds: 3
          restartPolicy: Always
          affinity:
            nodeAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                nodeSelectorTerms:
                  - matchExpressions:
                    - key: nodeType
                      operator: DoesNotExist
    ---
    apiVersion: v1
    kind: Service
    metadata:
      annotations:
        getambassador.io/config: |
          ---
          apiVersion: ambassador/v0
          kind:  Mapping
          name:  test-mapping
          prefix: /
          service: test-service
          host: "www.example.com"
          ambassador_id: ambassador-release
          timeout_ms: 30000
      name: test-service
      namespace: sandbox
    spec:
      ports:
      - port: 80
        protocol: TCP
        targetPort: 80
      selector:
        app: api
        slot: green
      sessionAffinity: None
      type: ClusterIP
    status:
      loadBalancer: {}
    ---
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: api-deployment-green
      namespace: sandbox
    spec:
      selector:
        matchLabels:
          app: api
          slot: green
      replicas: 1
      template:
        metadata:
          labels:
            app: api
            slot: green
        spec:
          containers:
          - name: httpd
            image: httpd:2.4.35
            ports:
            - containerPort: 80
    ---
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: api-deployment-blue
      namespace: sandbox
    spec:
      selector:
        matchLabels:
          app: api
          slot: blue
      replicas: 1
      template:
        metadata:
          labels:
            app: api
            slot: blue
        spec:
          containers:
          - name: nginx
            image: nginx:1.8
            ports:
            - containerPort: 80

**Comment by nbkrause** [comment]
Okay so you are changing the selector in `test-service` from `slot: green` to `slot: blue` and back and forth again. That clears some things up thank you.

So I was able to reproduce your issue yesterday but have been unable to today after deploying with helm. Yesterday, however, when I was able to reproduce I took a deeper look into what was going on and it appeared to be an issue with Kubernetes failing to point the service to the correct pod. 
**How I Tested This**
1. run `kubectl edit service test-service`
2. Test the edit through ambassador and notice the change didn't take place
3. Run `kubectl exec -it <ambassador-pod-name> -- /bin/sh` to open up a shell in the ambassador pod. This allows us to test within the cluster
4. Run curl `http://test-service/` and see that the wrong response is returned. This curl is bypassing ambassador and simply sending a request to `test-service` which is returning a response from whichever pod it is pointing to.

I am also skeptical of this being an issue with Ambassador. Ambassador simply takes this mapping annotation and sees to point requests to `http://<ambassador-ip>/` to `service: test-service`. `test-service` then is in charge of forwarding that request to the correct deployment. For this reason, and with the results of my testing, I am inclined to believe there is some Kubernetes error causing this.

I do believe I have a workaround though. While I was able to replicate this issue using `kubectl edit` I was unable to when I was editing the service using `kubectl apply -f`. The helm equivalent of this would be to edit the yaml template and `helm upgrade <release-name> <chart-directory>`.

I would be very interested in hearing your results of testing the traffic inside Kubernetes. I would also love to know if the workaround I suggested works.
